src/System.Memory/src/System/Buffers/SequenceReader.cs

```diff
@@ -0,0 +1,373 @@
+
+        /// <summary>
+        /// True when there is no more data in the <see cref="Sequence"/>.
+        /// </summary>
+        public bool End => !_moreData;
+
+        /// <summary>
+        /// The underlying <see cref="ReadOnlySequence{T}"/> for the reader.
+        /// </summary>
+        public ReadOnlySequence<T> Sequence { get; }
+
+        /// <summary>
+        /// The current position in the <see cref="Sequence"/>.
+        /// </summary>
+        public SequencePosition Position
+            => Sequence.GetPosition(CurrentSpanIndex, _currentPosition);
+
+        /// <summary>
+        /// The current segment in the <see cref="Sequence"/>.
+        /// </summary>
+        public ReadOnlySpan<T> CurrentSpan { get; private set; }
+
+        /// <summary>
+        /// The index in the <see cref="CurrentSpan"/>.
+        /// </summary>
+        public int CurrentSpanIndex { get; private set; }
+
+        /// <summary>
+        /// The unread portion of the <see cref="CurrentSpan"/>.
+        /// </summary>
+        public ReadOnlySpan<T> UnreadSpan
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => CurrentSpan.Slice(CurrentSpanIndex);
+        }
+
+        /// <summary>
+        /// The total number of <typeparamref name="T"/>'s processed by the reader.
+        /// </summary>
+        public long Consumed { get; private set; }
+
+        /// <summary>
+        /// Remaining <typeparamref name="T"/>'s in the reader's <see cref="Sequence"/>.
+        /// </summary>
+        public long Remaining => Length - Consumed;
+
+        /// <summary>
+        /// Count of <typeparamref name="T"/> in the reader's <see cref="Sequence"/>.
+        /// </summary>
+        public long Length
+        {
+            get
+            {
+                if (_length < 0)
+                {
+                    _length = Sequence.Length;
+                }
+                return _length;
+            }
+        }
+
+        /// <summary>
+        /// Peeks at the next value without advancing the reader.
+        /// </summary>
+        /// <param name="value">The next value or default if at the end.</param>
+        /// <returns>False if at the end of the reader.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool TryPeek(out T value)
+        {
+            if (_moreData)
+            {
+                value = CurrentSpan[CurrentSpanIndex];
+                return true;
+            }
+            else
+            {
+                value = default;
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Read the next value and advance the reader.
+        /// </summary>
+        /// <param name="value">The next value or default if at the end.</param>
+        /// <returns>False if at the end of the reader.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
There's a lot of AggressiveInlining happening here.  We believe that's the right trade-off?  It'll always look good in microbenchmarks.

Reply:
Tried not to do it where there wasn't a higher level measured impact or the code was called a lot.